mvpa2/tests/test_shrinklda.py

```diff
@@ -0,0 +1,74 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the PyMVPA package for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Unit tests for PyMVPA Shrinkage LDA classifier"""
+# based on test_gnb.py
+
+import numpy as np
+
+from mvpa2.clfs.skl.base import SKLLearnerAdapter
+
+from mvpa2.misc.data_generators import normal_feature_dataset
+from mvpa2.measures.base import CrossValidation
+from mvpa2.generators.partition import NFoldPartitioner
+from mvpa2.testing import *
+from mvpa2 import externals
+
+class ShrinkageLDATests(unittest.TestCase):
+
+    def test_shrinkage_lda_faster_than_skl(self):
+        # Compare PyMVPA's shrinkage LDA implementation with scikit-learn's
+        # implementation when used through SKLLearnerAdapter. Are prediction
+        # accuracies next to identical? Is the PyMVPA implementation faster?
+
+        # First check if scikit-learn is available at all
+        if not externals.exists('skl') or externals.versions['skl'] < '0.17':
+            self.skipTest('sklearn (0.17 or higher) unavailable')
+        else:
+            from mvpa2.clfs.shrinklda import ShrinkageLDA
+            from sklearn.discriminant_analysis import \
+            LinearDiscriminantAnalysis as sklLDA
+            # Dataset properties: should be of type "high p, low n" as is
+            # typical for fMRI data
+            n_features = 250
+            n_chunks = 3
+            n_classes = 3
+            n_samples_per_class = 18 # should be multiple of n_chunks
+            snr = 3.0
+
+            # Shrinkage estimators to use. OAS not an available option for
+            # sklLDA.
+            shrink_type = 'ledoit-wolf'
+
+            # The competitors
+            pymvpa_clf = ShrinkageLDA(shrinkage_estimator='ledoit-wolf')
+            # Using 'lsqr' because it is claimed to be more efficient than 'svd'
+            # see sklearn doc.
+            skl_clf = SKLLearnerAdapter(sklLDA(solver='lsqr', shrinkage='auto'))
+
+            ds = normal_feature_dataset(perlabel=n_samples_per_class,
+            nlabels=n_classes, nfeatures=n_features, nchunks=n_chunks,
+            nonbogus_features=range(n_classes), snr=snr)
+
+            partitioner = NFoldPartitioner(attr='chunks')
+
+            delta_t = []
+            for clf in [pymvpa_clf, skl_clf]:
+                cv = CrossValidation(clf, partitioner)
+                _ = cv(ds)
```

Comment:
well, test could compare those results -- your implementation should be not just faster but also providing a similar (better?) result, right? ;-)  otherwise -- what would be the point of having a faster but notably worse implementation?

Reply:
Thanks for the feedback!

1.) Yes, that's a cleaner solution with 'skip_if_no_external'.
2.) I though about this but I didn't know how to define the similarity of results. One idea I had was to show that the cv error of the new LDA implementation is more similar to the cv error from the old LDA version than from SVM classification. This is indeed the case in - say - 95 % of the cases. I figured this would not be acceptable for a test which would probably be run routinely together with a bunch of other tests. You don't want to risk the whole test failing just because you're in the last 5 % (or so) of this little test. Do you have a solution for this?

Otherwise I suggest that I elaborate a bit more on the documentation of the new ShrinkageLDA class to explain what it does. The code is quite simple and with a bit more explanation I could help the user convince himself/herself that what it does is indeed basic LDA with shrinkage on the within-class covariance matrix.